lib/autoproj/python.rb

```diff
@@ -0,0 +1,317 @@
+            ws.osdep_suffixes << "python#{$1}" if version =~ /^([0-9]+)\./
+
+            rewrite_python_shims(bin, ws.root_dir)
+            rewrite_pip_shims(bin, ws.root_dir)
+            [bin, version]
+        end
+
+        def self.deactivate_python(ws: Autoproj.workspace)
+            remove_python_shims(ws.root_dir)
+            remove_pip_shims(ws.root_dir)
+            ws.config.reset('python_executable')
+            ws.config.reset('python_version')
+        end
+
+        # Allow to update the PYTHONPATH for package if autoproj configuration
+        # USE_PYTHON is set to true.
+        # Then tries to guess the python binary from Autobuild.programs['python']
+        # and system's default setting
+        # @param [Autobuild::Package] pkg
+        # @param [Autoproj.workspace] ws Autoproj workspace
+        # @param [String] bin Path to a custom python version
+        # @param [String] version version constraint for python executable
+        # @return tuple of [executable, version, site-packages path] if set,
+        #    otherwise nil
+        def self.activate_python_path(pkg,
+                                 ws: Autoproj.workspace,
+                                 bin: nil,
+                                 version: nil)
+            return unless ws.config.get('USE_PYTHON', nil)
+
+            bin, version = resolve_python(ws: ws, bin: bin, version: version)
+            path = File.join(pkg.prefix, "lib",
+                             "python#{version}", "site-packages")
+            pkg.env_add_path 'PYTHONPATH', path
+
+            [bin, version, path]
+        end
+
+        def self.prepare_venv(ws: Autoproj.workspace, bin: nil, version: nil,
+                              venv_name: ".autoproj_python_venv")
+
+          python_bin, = resolve_python(ws: ws, bin: bin, version: version)
+          cmd = "#{python_bin} -m pip install --user -U virtualenv"
+          msg, status = Open3.capture2e(cmd)
+          unless status.success?
+              raise "Autoproj::Python.prepare_venv installation of "\
+                  " virtualenv failed: '#{msg}'"
+          end
+
+          cmd = "#{python_bin} -m virtualenv #{venv_name}"
+          msg, status = Open3.capture2e(cmd)
+          unless status.success?
+              raise "Autoproj::Python.prepare_venv preparation of"\
+                  " virtual env '#{venv_name}' failed: '#{msg}'"
+          end
+
+          ws.env.set('VIRTUAL_ENV_DISABLE_PROMPT', 1)
+          ws.env.source_after File.join(Autoproj.root_dir, venv_name, "bin", "activate")
+          File.join(Autoproj.root_dir, venv_name)
+        end
+
+        def self.setup_python_configuration_options(ws: Autoproj.workspace)
+            ws.config.declare 'USE_PYTHON', 'boolean',
+                              default: 'no',
+                              doc: ["Do you want to activate python?"]
+
+            if ws.config.get("USE_PYTHON")
+                ws.os_package_installer.install(%w[python pip])
```

Comment:
This is not necessary. `PipManager` [has an internal dependency on `pip` already](https://github.com/rock-core/autoproj/blob/master/lib/autoproj/package_managers/pip_manager.rb#L24)

Reply:
While I have overseen these given dependencies in PipManager, I wonder under what circumstances these dependencies are installed?

Reply:
They get installed when there's a package in the workspace that would be installed by the given package manager. All `PackageManagers` subclasses implement this method.

Reply:
Thanks, for pointing this out. I am not entirely sure, however, if this is a condition that is fulfilled for all users?

Reply:
Not sure I get your point. Pip should only be installed if the user needs pip (aka wants to install a package with pip), right?